core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+        blockedPercentMeters.put(name, KafkaMetricsGroup.newMeter(
+          s"${name}BlockedPercent", "blocked time", TimeUnit.NANOSECONDS, Map(ListenerMetricTag -> name)))
+    }
+  }
+
+  @After
+  def tearDown(): Unit = {
+    TestUtils.clearYammerMetrics()
+    blockedPercentMeters.clear()
+  }
+
+  @Test
+  def testFailWhenNoListeners(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    val executor = Executors.newSingleThreadExecutor
+    try {
+      // inc() on a separate thread in case it blocks
+      val externalListener = listeners("EXTERNAL")
+      executor.submit((() =>
+        intercept[RuntimeException](connectionQuotas.inc(externalListener.listenerName, externalListener.defaultClientIp, blockedPercentMeters("EXTERNAL")))): Runnable
+      ).get(5, TimeUnit.SECONDS)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testNoConnectionLimitsByDefault(): Unit = {
+    val config = KafkaConfig.fromProps(brokerPropsWithDefaultConnectionLimits)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // verify there is no limit by accepting 10000 connections as fast as possible
+      val numConnections = 10000
+      val futures = listeners.values.map( listener =>
+        executor.submit((() => acceptConnections(connectionQuotas, listener, numConnections)): Runnable) )
+      futures.foreach(_.get(10, TimeUnit.SECONDS))
+      listeners.values.foreach { listener =>
+        assertEquals(s"${listener.listenerName.value()}",numConnections, connectionQuotas.get(listener.defaultClientIp))
+
+        // verify removing one connection
+        connectionQuotas.dec(listener.listenerName, listener.defaultClientIp)
+        assertEquals(s"Number of connections on $listener:", numConnections - 1, connectionQuotas.get(listener.defaultClientIp))
+
+        // calling dec() for an IP for which we didn't call inc() should throw an exception
+        intercept[IllegalArgumentException](connectionQuotas.dec(listener.listenerName, unknownHost))
+      }
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testMaxConnectionsPerIp(): Unit = {
+    val maxConnectionsPerIp = 17
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionsPerIpProp, maxConnectionsPerIp.toString)
+    val config = KafkaConfig.fromProps(props)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    addListenersAndVerify(config, connectionQuotas)
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      val externalListener = listeners("EXTERNAL")
+      executor.submit((() => acceptConnections(connectionQuotas, externalListener, maxConnectionsPerIp)): Runnable).get(5, TimeUnit.SECONDS)
```

Comment:
What's the reason behind using an executor here? It seems that you always wait immediately `get` on the future thus we could directly do it in the main thread, isn't it? I would personally avoid them if not really necessary.

Reply:
the reason is to make sure that when `inc` blocks, it does not block the main thread where we cannot use a timeout. If it blocks on a separate thread, we make sure the test proceeds and fails (when it is not expected). Most of those are not supposed to block if the code works as we expect. I use the executor anyway, in case there is a bug or behavior changes and it unexpectedly blocks. In which case the whole test will hang for a long time (I guess till the build times out)

Reply:
Basically, the test does not assume that the code behaves correctly.